### DISCUSSION TITLE
Does Edge Cache persist between "cold starts" of serverless functions?

### DISCUSSION THREAD

**Discussion by mattjennings** [vercel/vercel#5193] [Help] [answered]
I'm using next.js deployed on vercel as a zero-config project. I have an API endpoint that will scrape a page and serve an image using puppeteer. It looks something like this:

```javascript
// /pages/api/i/[id].js
import chromium from 'chrome-aws-lambda'

export default async (req, res) => {
  const { id } = req.query
  const browser = await chromium.puppeteer.launch()
  const page = await browser.newPage()

  // load nextjs page
  await page.goto(`${process.env.BASE_URL}/v/${id}`, {
    timeout: 15 * 1000,
  })
  
  // take screenshot
  const data = await el.screenshot({
    type: 'png',
  })

  await browser.close()

  // return image
  res.setHeader('Cache-Control', 's-maxage=31536000, stale-while-revalidate')
  res.setHeader('Content-Type', 'image/png')
  res.end(data)
}
```

This function takes about 6~ seconds to respond on a cold-start, but after that it does return the cached response (the `x-vercel-cache` header in the response is `HIT`). However, when I try the next day with the same URL, it takes the same 6~ seconds and `x-vercel-cache` is `MISS`, despite it still being within the `s-maxage` seconds range. 

My best guess here is the cache will stay as long as the function is "warm"? If so, is there a way to persist it between cold starts?

**Comment by laneme** [comment]
hey did you find what was causing that issue?

**Reply by mattjennings** [reply]
Unfortunately I did not. I reached out to Vercel support in September and they started taking a look at it, but the last time I heard from them was mid-October. I’m going to reach out again soon.

**Comment by styfle** [comment, marked as answer]
> when I try the next day with the same URL, it takes the same 6~ seconds and x-vercel-cache is MISS

If you had no requests for 24 hours, then this is expected.

The cache already persists between cold starts. It is invalidated when you merge to the main branch and a new production deployment is created.

However, the edge cache is best effort and (as the name implies) per region. Since there are many cached responses in the region, the least recently used cached responses are invalidated to make room for more recent responses.

For these types of unusual apps (infrequent requests and high execution time) consider using external state, such as redis.

**Reply by mattjennings** [reply]
Ahh I see, that makes sense. Thanks for clarifying!

**Reply by dipunj** [reply]
Hey @styfle. I happen to have stumble across the same problem. I am building blog, which has a search api. I expect this api to be used rarely and for the end user it wouldn't be a good ux. Could you briefly describe how could i use redis with my nextjs blog for serving the search api.

My API is fairly basic, just a string as a GET query and then string matching a cached index file.

**Reply by styfle** [reply]
You could use [redis](https://www.npmjs.com/package/redis) as the client and [redislabs.com](https://redislabs.com) or [lambda.store](https://lambda.store) as your server.

Also read more about database solutions here: https://vercel.com/docs/solutions/databases

**Comment by f3l1x** [comment]
Thanks @styfle
